src/github.com/getlantern/enproxy/conn_test.go

```diff
@@ -172,34 +178,130 @@ func readResponse(conn net.Conn, req *http.Request, t *testing.T) {
+		destsReceived = make(map[string]bool)
+		statMutex.Unlock()
 		return
 	}
 
+	l, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		t.Fatalf("Proxy unable to listen: %v", err)
+	}
+	proxyAddr = l.Addr().String()
+
 	go func() {
 		proxy := &Proxy{
-			OnBytesReceived: func(clientIp string, bytes int64) {
-				atomic.AddInt64(&bytesReceived, bytes)
+			OnBytesReceived: func(clientIp string, destAddr string, req *http.Request, bytes int64) {
+				statMutex.Lock()
+				bytesReceived += bytes
+				destsReceived[destAddr] = true
+				statMutex.Unlock()
 			},
-			OnBytesSent: func(clientIp string, bytes int64) {
-				atomic.AddInt64(&bytesSent, bytes)
+			OnBytesSent: func(clientIp string, destAddr string, req *http.Request, bytes int64) {
+				statMutex.Lock()
+				bytesSent += bytes
+				destsSent[destAddr] = true
+				statMutex.Unlock()
 			},
 		}
-		err := proxy.ListenAndServe(PROXY_ADDR)
+		err := proxy.Serve(l)
+		if err != nil {
+			t.Fatalf("Proxy unable to serve: %s", err)
+		}
+	}()
+
+	if err := WaitForServer("tcp", proxyAddr, 1*time.Second); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func startHttpServer(t *testing.T) {
+	if httpAddr != "" {
+		return
+	}
+
+	l, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		t.Fatalf("HTTP unable to listen: %v", err)
+	}
+	httpAddr = l.Addr().String()
+
+	go func() {
+		httpServer := &http.Server{
+			Handler: http.HandlerFunc(func(resp http.ResponseWriter, req *http.Request) {
+				resp.Write([]byte(TEXT))
+			}),
+		}
+		err := httpServer.Serve(l)
 		if err != nil {
-			t.Fatalf("Unable to listen and serve: %s", err)
+			t.Fatalf("HTTP unable to serve: %s", err)
 		}
 	}()
-	if err := WaitForServer("tcp", PROXY_ADDR, 1*time.Second); err != nil {
+
+	if err := WaitForServer("tcp", httpAddr, 1*time.Second); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func startHttpsServer(t *testing.T) {
+	if httpsAddr != "" {
+		return
+	}
+
+	var err error
+
+	pk, err = keyman.GeneratePK(2048)
+	if err != nil {
+		t.Fatalf("Unable to generate key: %s", err)
+	}
+
+	// Generate self-signed certificate
+	cert, err = pk.TLSCertificateFor("tlsdialer", "localhost", time.Now().Add(1*time.Hour), true, nil)
+	if err != nil {
+		t.Fatalf("Unable to generate cert: %s", err)
+	}
+
+	keypair, err := tls.X509KeyPair(cert.PEMEncoded(), pk.PEMEncoded())
+	if err != nil {
+		t.Fatalf("Unable to generate x509 key pair: %s", err)
+	}
+
+	l, err := tls.Listen("tcp", "localhost:0", &tls.Config{
+		Certificates: []tls.Certificate{keypair},
+	})
+	if err != nil {
+		t.Fatalf("HTTP unable to listen: %v", err)
+	}
+	httpsAddr = l.Addr().String()
+
+	go func() {
+		httpServer := &http.Server{
+			Handler: http.HandlerFunc(func(resp http.ResponseWriter, req *http.Request) {
+				resp.Write([]byte(TEXT))
+			}),
+		}
+		err := httpServer.Serve(l)
+		if err != nil {
+			t.Fatalf("HTTPS unable to serve: %s", err)
+		}
+	}()
+
+	if err := WaitForServer("tcp", httpAddr, 1*time.Second); err != nil {
```

Comment:
startHttpServer() and startHttpsServer() share a significant amount of code. It might be worthwhile to put the last twenty lines of both in a separate, consolidated function and reference that to avoid some repetition?